text/0004-remove-unnecessary-allocator-usage.md

```diff
@@ -0,0 +1,326 @@
+# RFC-0004: Remove the host-side runtime memory allocator
+
+|                 |                                                                                             |
+| --------------- | ------------------------------------------------------------------------------------------- |
+| **Start Date**  | 2023-07-04                                                                                  |
+| **Description** | Update the runtime-host interface to no longer make use of a host-side allocator            |
+| **Authors**     | Pierre Krieger                                                                              |
+
+## Summary
+
+Update the runtime-host interface to no longer make use of a host-side allocator.
+
+## Motivation
+
+The heap allocation of the runtime is currently controlled by the host using a memory allocator on the host side.
+
+The API of many host functions consists in allocating a buffer. For example, when calling `ext_hashing_twox_256_version_1`, the host allocates a 32 bytes buffer using the host allocator, and returns a pointer to this buffer to the runtime. The runtime later has to call `ext_allocator_free_version_1` on this pointer in order to free the buffer.
+
+Even though no benchmark has been done, it is pretty obvious that this design is very inefficient. To continue with the example of `ext_hashing_twox_256_version_1`, it would be more efficient to instead write the output hash to a buffer that was allocated by the runtime on its stack and passed by pointer to the function. Allocating a buffer on the stack in the worst case scenario simply consists in decreasing a number, and in the best case scenario is free. Doing so would save many Wasm memory reads and writes by the allocator, and would save a function call to `ext_allocator_free_version_1`.
+
+Furthermore, the existence of the host-side allocator has become questionable over time. It is implemented in a very naive way, and for determinism and backwards compatibility reasons it needs to be implemented exactly identically in every client implementation. Runtimes make substantial use of heap memory allocations, and each allocation needs to go twice through the runtime <-> host boundary (once for allocating and once for freeing). Moving the allocator to the runtime side, while it would increase the size of the runtime, would be a good idea. But before the host-side allocator can be deprecated, all the host functions that make use of it need to be updated to not use it.
+
+## Stakeholders
+
+No attempt was made at convincing stakeholders.
+
+## Explanation
+
+### New host functions
+
+This section contains a list of new host functions to introduce.
+
+```wat
+(func $ext_storage_read_version_2
+    (param $key i64) (param $value_out i64) (param $offset i32) (result i64))
```

Comment:
> result could be an i32 here; even with an i32 a return value of -1 is not ambiguous as you'll physically never be able to actually allocate a buffer of such size.

I went with `i64` because I'm not really a fan of "`-1i32` is not ambiguous". If we ever switch to a 64bits runtime, then it's no longer impossible for 4 GiB to be allocated. Sure, a 4 GiB storage item would be unusually big, but it's also not in the real of completely absurd.

> correct?

Yes, but I think that your explanation builds upon a wrong assumption.

> if m == n (so either the buffer was exactly the right size or we still have some data left) then the guest enlarges the buffer and calls the host function again with a bigger offset and repeats this until the whole value is read (and if the guest guesses wrong then this could be called many times),

The idea when I designed this function is that the runtime is supposed to know what the size of the storage item is. When you're for example reading a `u32` from storage, you know it's always 4 bytes. When you're reading a `Vec<something>`, you can first read the length prefix in front of the `Vec` (which realistically shouldn't be more than 5 bytes) and then allocate a `Vec` of exactly the correct size.

Allocating a buffer whose size is equal to the size of the runtime storage item builds upon the assumption that you can't decode SCALE-encoded items in a streaming way. This assumption is true in practice now, but in the absolute it isn't.

To give an example, if you have a storage item whose format is `Vec<Vec<u8>>`, right now you have to allocate a big buffer that contains the entire storage item, then when decoding you copy the data within each inner `Vec`. It seems more optimal to me (but I could be wrong) to call `storage_read` once to get the size of the outer `Vec`, then twice for each inner `Vec`, and read the data of each inner `Vec<u8>`s directly its their destination buffer.